Extensions not stopped nor destroyed when add-on is uninstalled

**Describe the bug**
The extensions are not stopped nor destroyed when the corresponding add-on is uninstalled, preventing the extensions from properly terminate or clean up their resources.

**To Reproduce**
Steps to reproduce the behaviour:
1. Fuzz a message (e.g. send 10k empty payloads).
2. While the fuzzer is running uninstall the Fuzz add-on;
3. Note that in the log there are some errors (as the fuzzer extension was not destroyed).

**Expected behaviour**
The extension should be stopped and destroyed when the corresponding add-on is uninstalled.

**Software versions**
 - ZAP: 2.10.0

**Errors from the zap.log file**
```
475223 [ZAP-FuzzerThreadPool-0-thread-3] ERROR org.zaproxy.zap.utils.I18N - Failed to load a message:
java.util.MissingResourceException: Can't find resource for bundle java.util.PropertyResourceBundle, key fuzz.httpfuzzer.messagetype.result
	at java.util.ResourceBundle.getObject(ResourceBundle.java:564) ~[?:?]
	at java.util.ResourceBundle.getString(ResourceBundle.java:521) ~[?:?]
	at org.zaproxy.zap.utils.I18N.getStringImpl(I18N.java:185) [zap-2.10.0.jar:2.10.0]
	at org.zaproxy.zap.utils.I18N.getString(I18N.java:149) [zap-2.10.0.jar:2.10.0]
	at org.zaproxy.zap.extension.fuzz.httpfuzzer.HttpFuzzerTask.runImpl(HttpFuzzerTask.java:59) [fuzz-beta-13.1.0.zap:?]
	at org.zaproxy.zap.extension.fuzz.httpfuzzer.HttpFuzzerTask.runImpl(HttpFuzzerTask.java:30) [fuzz-beta-13.1.0.zap:?]
	at org.zaproxy.zap.extension.fuzz.AbstractFuzzerTask.run(AbstractFuzzerTask.java:56) [fuzz-beta-13.1.0.zap:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
501048 [ZAP-FuzzerThreadPool-0-thread-5] ERROR org.zaproxy.zap.ZAP.UncaughtExceptionLogger - Exception in thread "ZAP-FuzzerThreadPool-0-thread-5"
java.lang.NoClassDefFoundError: org/zaproxy/zap/extension/fuzz/AbstractFuzzer$ExecutorTerminatedListenerImpl$1
	at org.zaproxy.zap.extension.fuzz.AbstractFuzzer$ExecutorTerminatedListenerImpl.terminated(AbstractFuzzer.java:664) ~[?:?]
	at org.zaproxy.zap.utils.PausableThreadPoolExecutor.notifyListeners(PausableThreadPoolExecutor.java:276) ~[zap-2.10.0.jar:2.10.0]
	at org.zaproxy.zap.utils.PausableThreadPoolExecutor.terminated(PausableThreadPoolExecutor.java:266) ~[zap-2.10.0.jar:2.10.0]
	at java.util.concurrent.ThreadPoolExecutor.tryTerminate(ThreadPoolExecutor.java:718) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.processWorkerExit(ThreadPoolExecutor.java:994) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: java.lang.ClassNotFoundException
	at org.zaproxy.zap.control.AddOnClassLoader.findClass(AddOnClassLoader.java:330) ~[zap-2.10.0.jar:2.10.0]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:589) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?]
	... 8 more
```

--- a/zap/src/main/java/org/parosproxy/paros/extension/Extension.java
+++ b/zap/src/main/java/org/parosproxy/paros/extension/Extension.java
@@ -46,6 +46,7 @@
 // ZAP: 2019/06/05 Normalise format/style.
 // ZAP: 2019/09/12 Add defaults to getAuthor() and getURL().
 // ZAP: 2020/03/09 Default getURL() to null.
+// ZAP: 2021/04/13 Update JavaDoc of stop() and destroy().
 package org.parosproxy.paros.extension;
 
 import java.net.URL;
@@ -127,11 +128,27 @@ public interface Extension {
     /** Start the plugin e.g. if there is a running server. */
     void start();
 
-    /** stop the plugin e.g. if there is a running server. */
+    /**
+     * Stops the extension, e.g. running scans, processes, servers.
+     *
+     * <p>Called when the extension is removed (i.e. corresponding add-on is uninstalled) and when
+     * ZAP shuts down.
+     *
+     * <p>Should be called only by core functionality.
+     *
+     * @see #destroy()
+     */
     void stop();
 
     /**
-     * Plugin cleanup, finalize, etc. when program shutdown. Stop() will be called before shutdown.
+     * Performs final cleanups, free resources.
+     *
+     * <p>Called when the extension is removed (i.e. corresponding add-on is uninstalled) and when
+     * ZAP shuts down.
+     *
+     * <p>Called after {@link #stop() stopping} the extension.
+     *
+     * <p>Should be called only by core functionality.
      */
     void destroy();
 

--- a/zap/src/main/java/org/parosproxy/paros/extension/ExtensionLoader.java
+++ b/zap/src/main/java/org/parosproxy/paros/extension/ExtensionLoader.java
@@ -89,6 +89,7 @@
 // ZAP: 2020/05/14 Hook HttpSenderListener when starting single extension.
 // ZAP: 2020/08/27 Added support for plugable variants
 // ZAP: 2020/11/26 Use Log4j 2 classes for logging.
+// ZAP: 2021/04/13 Issue 6536: Stop and destroy extensions being removed.
 package org.parosproxy.paros.extension;
 
 import java.awt.Component;
@@ -1473,6 +1474,9 @@ public class ExtensionLoader {
     /**
      * Removes the given extension and any components added through its extension hook.
      *
+     * <p>The extension is also {@link Extension#stop() stopped} and {@link Extension#destroy()
+     * destroyed}.
+     *
      * <p><strong>Note:</strong> This method should be called only by bootstrap classes.
      *
      * @param extension the extension to remove.
@@ -1482,6 +1486,12 @@ public class ExtensionLoader {
         extensionList.remove(extension);
         extensionsMap.remove(extension.getClass());
 
+        extension.stop();
+        unhook(extension);
+        extension.destroy();
+    }
+
+    private void unhook(Extension extension) {
         ExtensionHook hook = extensionHooks.remove(extension);
         if (hook == null) {
             logger.error("ExtensionHook not found for: " + extension.getClass().getCanonicalName());

--- a/zap/src/test/java/org/parosproxy/paros/extension/ExtensionLoaderUnitTest.java
+++ b/zap/src/test/java/org/parosproxy/paros/extension/ExtensionLoaderUnitTest.java
@@ -1,0 +1,77 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2021 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.parosproxy.paros.extension;
+
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InOrder;
+import org.parosproxy.paros.model.Model;
+
+/** Unit test for {@link ExtensionLoader}. */
+class ExtensionLoaderUnitTest {
+
+    private Model model;
+    private ExtensionLoader extensionLoader;
+
+    @BeforeEach
+    void setUp() {
+        model = mock(Model.class);
+        extensionLoader = new ExtensionLoader(model, null);
+    }
+
+    @Test
+    void shouldStopExtensionWhenRemoved() {
+        // Given
+        Extension extension = mock(Extension.class);
+        extensionLoader.addExtension(extension);
+        // When
+        extensionLoader.removeExtension(extension);
+        // Then
+        verify(extension).stop();
+    }
+
+    @Test
+    void shouldDestroyExtensionWhenRemoved() {
+        // Given
+        Extension extension = mock(Extension.class);
+        extensionLoader.addExtension(extension);
+        // When
+        extensionLoader.removeExtension(extension);
+        // Then
+        verify(extension).destroy();
+    }
+
+    @Test
+    void shouldStopBeforeDestroyExtensionWhenRemoved() {
+        // Given
+        Extension extension = mock(Extension.class);
+        extensionLoader.addExtension(extension);
+        InOrder inOrder = inOrder(extension);
+        // When
+        extensionLoader.removeExtension(extension);
+        // Then
+        inOrder.verify(extension).stop();
+        inOrder.verify(extension).destroy();
+    }
+}

